src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -156,32 +178,114 @@ public synchronized void update(Iterable<String> inputs) {
     while (header.before == header.after) {
       wait();
     }
+    // should assert that e.referenceCount == 0
     Entry e = header.after;
     if (e.referenceCount != 0) {
       throw new RuntimeException("ERROR: Reference counts lru ordering has not been maintained correctly, attempting to expire referenced (or negatively counted) content");
     }
     storage.remove(e.key);
+    for (Digest containingDirectory : e.containingDirectories) {
+      expireDirectory(containingDirectory);
+    }
     e.remove();
     sizeInBytes -= e.size;
     return e.key;
   }
 
+  /** must be called in synchronized context */
+  private void expireDirectory(Digest digest) throws IOException {
+    DirectoryEntry e = directoryStorage.remove(digest);
+    Path path = getDirectoryPath(digest);
+
+    for (String input : e.inputs) {
+      Entry fileEntry = storage.get(input);
+
+      if (fileEntry != null) {
+        fileEntry.containingDirectories.remove(digest);
+      }
+    }
+    Worker.removeDirectory(path);
+  }
+
+  /** must be called in synchronized context */
+  private void incrementReferences(Iterable<String> inputs) {
+    for (String input : inputs) {
+      storage.get(input).incrementReference(header);
+    }
+  }
+
+  /** must be called in synchronized context */
+  private void fetchDirectory(
+      Digest containingDirectory,
+      Path path,
+      Digest digest,
+      Map<Digest, Directory> directoriesIndex,
+      ImmutableList.Builder<String> inputsBuilder) throws IOException, InterruptedException {
+    Directory directory = directoriesIndex.get(digest);
+    Files.createDirectory(path);
+    for (FileNode fileNode : directory.getFilesList()) {
+      String fileCacheKey = put(fileNode.getDigest(), fileNode.getIsExecutable(), containingDirectory);
+      Files.createLink(path.resolve(fileNode.getName()), getPath(fileCacheKey));
+      inputsBuilder.add(fileCacheKey);
+    }
+    for (DirectoryNode directoryNode : directory.getDirectoriesList()) {
+      fetchDirectory(
+          containingDirectory,
+          path.resolve(directoryNode.getName()),
+          directoryNode.getDigest(),
+          directoriesIndex,
+          inputsBuilder);
+    }
+  }
+
+  public synchronized Path putDirectory(
+      Digest digest,
+      Map<Digest, Directory> directoriesIndex) throws IOException, InterruptedException {
+    Path path = getDirectoryPath(digest);
+
+    DirectoryEntry e = directoryStorage.get(digest);
+    if (e != null) {
+      incrementReferences(e.inputs);
+      return path;
+    }
+
+    ImmutableList.Builder<String> inputsBuilder = new ImmutableList.Builder<>();
+    fetchDirectory(digest, path, digest, directoriesIndex, inputsBuilder);
+
+    e = new DirectoryEntry(directoriesIndex.get(digest), inputsBuilder.build());
+
+    directoryStorage.put(digest, e);
+
+    return path;
+  }
+
   public String put(Digest digest, boolean isExecutable) throws InterruptedException {
```

Comment:
I'm curious why you sometimes use String and other times Path as a result of put. IIUC, this is always the local "true" path of the file or directory, right?

Reply:
I think that was just a result of trying to avoid creating additional objects, but it is likely moot or better (certainly simplifies some things) when compared to storing Paths. I just tried it out here, seems legit - posting change